compiler/noirc_evaluator/src/ssa/opt/remove_bit_shifts.rs

```diff
@@ -222,39 +223,96 @@ impl Context<'_, '_, '_> {
                 // unsigned right bit shift is just a normal division
                 self.insert_binary(lhs, BinaryOp::Div, pow)
             }
-            NumericType::Signed { bit_size } => {
+            NumericType::Signed { .. } => {
+                let mut instruction_builder = InstructionBuilder::new(
+                    self.context.dfg,
+                    self.context.block_id,
+                    self.context.call_stack_id,
+                );
+
+                let zero = instruction_builder.numeric_constant(FieldElement::zero(), lhs_typ);
+                let lhs_sign = instruction_builder.insert_binary(lhs, BinaryOp::Lt, zero);
+
                 // Get the sign of the operand; positive signed operand will just do a division as well
-                let zero =
-                    self.numeric_constant(FieldElement::zero(), NumericType::signed(bit_size));
-                let lhs_sign = self.insert_binary(lhs, BinaryOp::Lt, zero);
-                let lhs_sign_as_field = self.insert_cast(lhs_sign, NumericType::NativeField);
-                let lhs_as_field = self.insert_cast(lhs, NumericType::NativeField);
-                // For negative numbers, convert to 1-complement using wrapping addition of a + 1
-                // Unchecked add as these are fields
-                let add = BinaryOp::Add { unchecked: true };
-                let one_complement = self.insert_binary(lhs_sign_as_field, add, lhs_as_field);
-                let one_complement = self.insert_truncate(one_complement, bit_size, bit_size + 1);
                 let one_complement =
-                    self.insert_cast(one_complement, NumericType::signed(bit_size));
+                    instruction_builder.convert_twos_complement_to_ones_complement(lhs, lhs_sign);
```

Comment:
FYI I updated https://github.com/noir-lang/noir/pull/9965/files#diff-6a664951cab26f1ffa0e8854826acb88fe48bc294100187283ebd81cd6a7c874 with more comments after discussing with Guillaume: the comments that said we are converting from 2-complement to 1-complement were misleading, that's not what it did. I don't think there's any particular name for the converted value.

Reply:
Thanks for flagging this up! I'll take a look and update this PR